samcli/lib/pipeline/bootstrap/stage.py

```diff
@@ -0,0 +1,280 @@
+
+        THIS METHOD UPDATES THE STATE OF THE CALLING INSTANCE(self) IT WILL SET THE VALUES OF THE RESOURCES ATTRIBUTES
+
+        Parameters
+        ----------
+        confirm_changeset: bool
+            if set to false, the stage_resources.yaml CFN template will directly be deployed, otherwise, the user will
+            be prompted for confirmation
+
+        Returns True if bootstrapped, otherwise False
+        """
+
+        if self.did_user_provide_all_required_resources():
+            click.secho(f"\nAll required resources for the {self.name} stage exist, skipping creation.", fg="yellow")
+            return True
+
+        missing_resources_msg: str = self._get_non_user_provided_resources_msg()
+        click.echo(
+            "This will create the following required resources for the {self.name} stage: " f"{missing_resources_msg}"
+        )
+        if confirm_changeset:
+            confirmed: bool = click.confirm("Should we proceed with the creation?")
+            if not confirmed:
+                return False
+
+        stage_name_suffix: str = re.sub("[^0-9a-zA-Z]+", "-", self.name)
+        stack_name: str = f"{STACK_NAME_PREFIX}-{stage_name_suffix}-{STAGE_RESOURCES_STACK_NAME_SUFFIX}"
+        stage_resources_template_body = Stage._read_template(STAGE_RESOURCES_CFN_TEMPLATE)
+        output: StackOutput = manage_stack(
+            stack_name=stack_name,
+            region=self.aws_region,
+            profile=self.aws_profile,
+            template_body=stage_resources_template_body,
+            parameter_overrides={
+                "PipelineUserArn": self.pipeline_user.arn or "",
+                "PipelineExecutionRoleArn": self.pipeline_execution_role.arn or "",
+                "PipelineIpRange": self.pipeline_ip_range or "",
+                "CloudFormationExecutionRoleArn": self.cloudformation_execution_role.arn or "",
+                "ArtifactsBucketArn": self.artifacts_bucket.arn or "",
+                "CreateEcrRepo": "true" if self.create_ecr_repo else "false",
+                "EcrRepoArn": self.ecr_repo.arn or "",
+            },
+        )
+
+        self.pipeline_user.arn = output.get("PipelineUser")
+        self.pipeline_user.access_key_id = output.get("PipelineUserAccessKeyId")
+        self.pipeline_user.secret_access_key = output.get("PipelineUserSecretAccessKey")
+        self.pipeline_execution_role.arn = output.get("PipelineExecutionRole")
+        self.cloudformation_execution_role.arn = output.get("CloudFormationExecutionRole")
+        self.artifacts_bucket.arn = output.get("ArtifactsBucket")
+        self.artifacts_bucket.kms_key_arn = output.get("ArtifactsBucketKMS")
+        self.ecr_repo.arn = output.get("EcrRepo")
+        return True
+
+    @staticmethod
+    def _read_template(template_file_name: str) -> str:
+        template_path: str = os.path.join(CFN_TEMPLATE_PATH, template_file_name)
+        with open(template_path, "r") as fp:
```

Comment:
This will fail on Windows on UTF-8 bytes. See e.g. https://github.com/aws/aws-sam-cli/pull/2583

I'm surprised we don't have a generic method for reading a file as a string.

Reply:
Are you saying python's builtin `open` function working on windows? the issue you are linking uses `click.open`

Reply:
It uses `open` under the hood. If `encoding` isn't specified, it defaults to whatever's the system default (typically Latin-1 on Windows).